examples/acoustic/fwi_operators.py

```diff
@@ -34,25 +34,28 @@ def ForwardOperator(model, u, src, rec, data, time_order=2, spc_order=6,
         # eqn = m * u.dt2 - laplacian - s**2 / 12 * biharmonic + damp * u.dt
         dt = 1.73 * model.critical_dt
 
+    # Derive both stencils from symbolic equation:
     # Create the stencil by hand instead of calling numpy solve for speed purposes
     # Simple linear solve of a u(t+dt) + b u(t) + c u(t-dt) = L for u(t+dt)
     stencil = 1 / (2 * m + s * damp) * (
         4 * m * u + (s * damp - 2 * m) * u.backward +
         2 * s**2 * (laplacian + s**2 / 12 * biharmonic))
-    # Add substitutions for spacing (temporal and spatial)
-    subs = {s: dt, h: model.get_spacing()}
+    eqn = Eq(u.forward, stencil)
 
-    dse = kwargs.get('dse', 'advanced')
-    dle = kwargs.get('dle', 'advanced')
+    # Construct expression to inject source values
+    # Note that src and field terms have differing time indices:
+    #   src[time, ...] - always accesses the "unrolled" time index
+    #   u[ti + 1, ...] - accesses the forward stencil value
+    ti = u.indices[0]
+    source = src.inject(field=u.indexed[ti + 1, ...], offset=model.nbpml,
+                        expr=src.indexed[time, ...] * dt * dt / m)
 
```

Comment:
Ok, so first thing to note is that the previous `point2grid` function call was simply hiding some very ugly notation that derived the correct indices by hand and hard-codes the expression `src * dt * dt / m` - therefore it cannot really be generalised (see `source_type.py:99-109`). This is in contrast is raw **Devito low-level notation** (ie. our backdoor), as indicated by the `indexed` keyword.

The reason I can only provide the ellipsis shorthand in low-level notation (and believe me I tried doing it on the high-level!), is simply because over-writing `__getitem__` on a `sympy.Function` causes a multitude of low-level SymPy failures in things like `sympy.solve`, because we all of a sudden provide an iteration protocol, although `Function` objects are not actually iterable. I tried things like raising a failure through `__iter__`, but alas no luck. If someone can provide another idea on how to prevent `Function` iteration, we could move this trick to the high-level, so further suggestions are welcome...